Add app description

IvanKiral opened the pull request:
### Motivation

Update readme to describe how the application works

### Checklist

- [ ] Code follows coding conventions held in this repo
- [ ] Automated tests have been added
- [ ] Tests are passing
- [ ] Docs have been updated (if applicable)
- [ ] Temporary settings (e.g. variables used during development and testing) have been reverted to defaults

### How to test

If manual testing is required, what are the steps?

Simply007 reviewed (changes requested):
Text is good.

I would think, how to adjust the Model Mapping and Data Fetching for smoother flow.

I would love to have in every section some code sample in pseudo typescript. I have drafted some, so try to apply this.

Simply007 reviewed (changes requested):
There is already one section about deployment

+ rut lint on readme

IvanKiral commented on README.md:
````diff
@@ -76,6 +76,50 @@ For more info about the API, see the [API reference](https://kontent.ai/learn/re
+
+```{typescript}
+const query = Client.items<generatedDTO>()
+      .type(projectModel.contentTypes.generatedDTO.codename)...
````
Decided to go with GeneratedDTO as all generated models starts with uppercase

IvanKiral commented on README.md:
````diff
@@ -76,6 +76,50 @@ For more info about the API, see the [API reference](https://kontent.ai/learn/re
+const query = Client.items<generatedDTO>()
+      .type(projectModel.contentTypes.generatedDTO.codename)...
+```
+
````
I think there might be a little misunderstanding. I do not think that we use stateless components in pages. Pages fetch data by themselves. Only example which I could find to use components pattern are `BrewerStoreContainer` with `BrewerStoreListing` and `CoffeStoreContainer` with `CoffeStoreListing`

Are those what you have in mind? Because comment below you have code sample of page which fetch data by itself.

IvanKiral commented:
I have merged data fetching and content modelling into one section, however it seemed to me better to have only one code block there, as it does not make much sense to showcase how the interface looks like. On the other hand as you said people mostly find the the code and then read the lines above... Should I somehow add some generic interface for type there?

From other changes, I made some paragraphs shorter, as I felt that some information were unnecessary. I removed some links, because I thought that it may annoy people more, when there is a lot of links to other files. Also I run prettier on the file. And lastly I have added code blocks for more paragraphs. For instance for Handling 404 I did not add any, it only explain concept what we did, not what developer can do in the project.

Simply007 commented on README.md:
```diff
@@ -76,6 +76,163 @@ For more info about the API, see the [API reference](https://kontent.ai/learn/re
 
+## About
+
+This section describes the application.
```
We should either summarize the content of the section in this part, or remove this sentence - it is ambignous

Simply007 commented on README.md:
```diff
@@ -76,6 +76,163 @@ For more info about the API, see the [API reference](https://kontent.ai/learn/re
+
+#### content type -> DTO -> view model -> component
+
+Some models displayed in views might require a small adjustment from content types. For example, `Cafe` content type contains fields for `city` and `street` and we would like to have model containing an address in the format `city, street`. You can find an example for such a view model in `CafeModel.tsx` that can be found in the `src/ViewModels` folder. To convert `Cafe` into `CafeModel` you can  ...
```
```suggestion
Some models displayed in views might require an adjustment from content types. For example, the `Cafe` content type contains fields for `city` and `street` and we would like to have a model containing an address in the format `city, street`. You can find an example for such a view model in `CafeModel.tsx` that can be found in the `src/ViewModels` folder. To convert `Cafe` into `CafeModel` you can use the function located in `src/Utilities/CafeListing.ts`

```

Simply007 commented on README.md:
```diff
@@ -76,6 +76,163 @@ For more info about the API, see the [API reference](https://kontent.ai/learn/re
+
+#### content type -> DTO -> component
+
+For generating component models from content types, we have used [Kontent.ai model generator](https://github.com/Kentico/kontent-model-generator-js) tool. All generated models can be found in `src/Models` folder. The `_project.ts` exports `projectModel` which contains information about the project structure such as project languages as well as other information such as codenames about content typ ...
```
```suggestion
Content type definitions are being generated from content types via [Kontent.ai model generator](https://github.com/Kentico/kontent-model-generator-js) tool. All generated types can be found in `src/Models` folder. The `_project.ts` contains information about the project structure such as project languages as well as other structure information such as codenames about content types.
```

We might want to adjust the naming - it is not a model anymore - it is a type... soryy I might have point that out sooner.

Simply007 commented on README.md:
```diff
@@ -76,6 +76,163 @@ For more info about the API, see the [API reference](https://kontent.ai/learn/re
+
+#### Data fetching
+
+This solution fetches data using the [Delivery client](https://github.com/Kentico/kontent-delivery-sdk-js). For more implementation detail to setup client see `src/Client.ts`. The data are fetched and stored in a `container` component and they are then passed to `presentation` component. For better understanding see the code example below. However, depending on your needs, you can use other techn ...
```
```suggestion
This solution fetches data using the [Delivery client](https://github.com/Kentico/kontent-delivery-sdk-js). For more implementation detail to set up the client see `src/Client.ts`. The data are fetched and stored in a `container` component directly in its state. Then they are passed to the `presentation` component. For a better understanding see the code example below. However, depending on your needs, you can use other technologies for managing application states such as:
```

Simply007 commented on README.md:
```diff
@@ -76,6 +76,163 @@ For more info about the API, see the [API reference](https://kontent.ai/learn/re
+
+This solution fetches data using the [Delivery client](https://github.com/Kentico/kontent-delivery-sdk-js). For more implementation detail to setup client see `src/Client.ts`. The data are fetched and stored in a `container` component and they are then passed to `presentation` component. For better understanding see the code example below. However, depending on your needs, you can use other techn ...
+
+- Context
```
We might want to add links to these :)

Simply007 commented on README.md:
````diff
@@ -76,6 +76,163 @@ For more info about the API, see the [API reference](https://kontent.ai/learn/re
+- Flux
+- ...
+
+```typescript
````
```suggestion
```tsx
```

Simply007 commented on README.md:
````diff
@@ -76,6 +76,163 @@ For more info about the API, see the [API reference](https://kontent.ai/learn/re
+}
+```
+
+### Filtering in product catalog
````
```suggestion
### Filter by taxonomy
```

Simply007 commented on README.md:
````diff
@@ -76,6 +76,163 @@ For more info about the API, see the [API reference](https://kontent.ai/learn/re
+
+Filters in Kontent.ai are implemented using taxonomies. Filtering examples can be found in `src/Components/BrewerStoreContainer.tsx` or `src/Components/CoffeeStoreContainer.tsx`. Firstly, the taxonomies groups that contain possible values for filters are loaded in `useEffect` blocks. We store selected values for filtering in `filter` variable. Items to be displayed are then selected with function ...
+
+```tsx
````
```suggestion
```tsx
```

I might also use proper naming: `processing` and `productStatuses` - filterVariable1 and filerVafiable2 is hard to read

Simply007 commented on README.md:
```diff
@@ -76,6 +76,163 @@ For more info about the API, see the [API reference](https://kontent.ai/learn/re
+
+#### Resoure strings
+
+Not every text of the application must be stored in Kontent.ai. Some strings, such as button texts, navigation texts and so on, can be stored locally. For those texts we use [React-intl](https://formatjs.io/docs/getting-started/installation/). For every language we have created JSON file in `src/Localization` folder. As we use `React-intl` it can not parse nested JSON objects and therefore the fo ...
```
```suggestion
Not every text of the application must be stored in Kontent.ai. Some strings, such as button texts, navigation texts, and so on, can be stored directly in the application. For those texts we use [React-intl](https://formatjs.io/docs/getting-started/installation/). For every language, there is a JSON file in `src/Localization` folder. 

> `React-intl` can not parse nested JSON objects and therefore the format of files is `key:value`. To load all files from `src/Localization` folder there is a `src/utilities/LocalizationLoader.ts` script.
```

Simply007 commented on README.md:
````diff
@@ -76,6 +76,163 @@ For more info about the API, see the [API reference](https://kontent.ai/learn/re
+}
+```
+
+#### Prefixes and Localizable Url slugs
````
```suggestion
#### Language prefixes
```

Simply007 commented on README.md:
````diff
@@ -76,6 +76,163 @@ For more info about the API, see the [API reference](https://kontent.ai/learn/re
+    }
+    ...
+```
+
````
```suggestion

#### Localizable URL slugs
```

Simply007 commented on README.md:
````diff
@@ -76,6 +76,163 @@ For more info about the API, see the [API reference](https://kontent.ai/learn/re
+```
+
+You might want to request items based on the url slugs. For more information how it works in Kontent see this [link](https://kontent.ai/learn/tutorials/develop-apps/get-content/localized-content-items/#a-get-items-by-localized-url-slug). We have provided an example in this application for `src/Pages/About.tsx' page.
+
````
```suggestion

> The showcase is not ideal, because it is using a combination of the language prefix and localizable solution is not ideal. You should try to stick with one of the approaches. Because it is hard to define the behavior for language setting clash i.e. `/<EN-PREFIX>/articles/<ES-URL-SLUG>`.
```

Simply007 commented on README.md:
````diff
@@ -76,6 +76,163 @@ For more info about the API, see the [API reference](https://kontent.ai/learn/re
+    ...
+```
+
+You might want to request items based on the url slugs. For more information how it works in Kontent see this [link](https://kontent.ai/learn/tutorials/develop-apps/get-content/localized-content-items/#a-get-items-by-localized-url-slug). We have provided an example in this application for `src/Pages/About.tsx' page.
````
```suggestion
You might want to request items based on the URL slugs. For more information check out [Kontent.ai/learn tutorial](https://kontent.ai/learn/tutorials/develop-apps/get-content/localized-content-items/#a-get-items-by-localized-url-slug). We have provided an example in this application for `src/Pages/About.tsx` page.

```

Simply007 reviewed (changes requested):
For further reference - the best form for readme is to use passive. If you are up for it - you can rewrite the text in this form, but let's keep it optional - for important parts - I have placed the suggestions.

Before sending to another review - run the text using Grammarly.

Simply007 commented on README.md:
````diff
@@ -76,6 +76,165 @@ For more info about the API, see the [API reference](https://kontent.ai/learn/re
+
+```tsx
+const Component: React.FC = () => {
+    const [data, setData] = useState([] : DTOComponentProps);
````
```suggestion
    const [data, setData] = useState<GeneratedDTO[]>([]);

```

Simply007 commented on README.md:
```diff
@@ -76,6 +76,165 @@ For more info about the API, see the [API reference](https://kontent.ai/learn/re
+          .type(projectModel.contentTypes.generatedDTO.codename)
+          ...
+
+      const items : GeneratedDTO[] = query.ToPromise()
```
There should not be an assignment - `setState` does not return anything: https://reactjs.org/docs/react-component.html#setstate

Simply007 commented on README.md:
```diff
@@ -76,6 +76,165 @@ For more info about the API, see the [API reference](https://kontent.ai/learn/re
+          ...
+
+      const items : GeneratedDTO[] = query.ToPromise()
+          .then(data => data.items()
```
```suggestion
          .then(data => data.items)
```

Simply007 replied:
We might want to merge the two the function, but I will keep the decision on you

Simply007 commented on README.md:
```diff
@@ -76,6 +76,165 @@ For more info about the API, see the [API reference](https://kontent.ai/learn/re
+
+In Kontent each language is identified by codename, in case of this project, it is `en-US` and `es-ES`.
+
+### Resoure strings
```
```suggestion
### Resource strings

```

Simply007 reviewed (changes requested):
A couple of nitpicks

Simply007 reviewed (changes requested):
One last thing
